cache.go

```diff
@@ -0,0 +1,872 @@
+					return
+				}
+				if statusCode == http.StatusNotModified {
+					// add end to headers to cache response headers
+					// serve from cache
+					serveCache = true
+					endHdrs := getEndToEndHeaders(result.Header)
+					for k, v := range endHdrs {
+						cacheHdrs.Header[k] = v
+					}
+				} else if statusCode != http.StatusOK {
+					go func() {
+						clnt.RemoveObject(clnt.bucket, key)
+					}()
+					// write backend response and return
+				}
+			}
+			if serveCache {
+				// serve cache entry
+				setRespHeaders(w, cacheHdrs.Header)
+				statusCodeWritten := false
+				if cacheHdrs.Range() != nil {
+					w.WriteHeader(http.StatusPartialContent)
+				}
+				httpWriter := xioutil.WriteOnClose(w)
+				// Write object content to response body
+				if _, err := io.Copy(httpWriter, reader); err != nil {
+					if !httpWriter.HasWritten() && !statusCodeWritten { // write error response only if no data or headers has been written to client yet
+						b.proxy.ServeHTTP(w, r)
+					}
+					return
+				}
+
+				if err := httpWriter.Close(); err != nil {
+					if !httpWriter.HasWritten() && !statusCodeWritten { // write error response only if no data or headers has been written to client yet
+						b.proxy.ServeHTTP(w, r)
+						return
+					}
+				}
+				return
+			}
+			for k, v := range result.Header {
+				w.Header().Set(k, strings.Join(v, ","))
+			}
+
+			mustCache := true
+			resultContentLength := result.ContentLength
+			respCC := parseCacheControlHeaders(result.Header)
+			if respCC.neverCache() || resultContentLength < clnt.minSize {
+				if cacheErr == nil {
+					go func() {
+						clnt.RemoveObject(clnt.bucket, key)
+					}()
+				}
+				mustCache = false
+			}
+			rs := result.Header.Get(xhttp.ContentRange)
+			if rs != "" {
+				// Avoid caching range GET's for now.
+				if cacheErr == nil {
+					go func() {
+						clnt.RemoveObject(clnt.bucket, key)
+					}()
+				}
+				mustCache = false
+			}
+			var pipeReader *io.PipeReader
+			var pipeWriter *io.PipeWriter
+			var mw io.Writer
+			// default to http.ResponseWriter
+			mw = w
+			// overwrite mw with MultiWriter if caching
+			if result.ContentLength >= clnt.minSize && mustCache {
+				pipeReader, pipeWriter = io.Pipe()
+				mw = io.MultiWriter(w, pipeWriter)
+			}
+			statusCode := result.StatusCode
+			w.WriteHeader(statusCode)
+
+			if mustCache && resultContentLength >= clnt.minSize {
+				go func() {
+					_, err := clnt.PutObject(clnt.bucket, key,
+						io.LimitReader(pipeReader, resultContentLength), resultContentLength,
+						"", "", getPutMetadata(result.Header), nil)
+					if err != nil {
+						clnt.setOffline()
+						logger.LogIf(context.Background(), err, "Failed to cache object")
+					}
+					pipeReader.CloseWithError(err)
+				}()
+			}
+			_, err := io.Copy(mw, io.LimitReader(result.Body, result.ContentLength))
+			if err != nil {
+				logger.LogIf(context.Background(), err)
+				pipeWriter.CloseWithError(err)
```

Comment:
you can move this outside of this `if` it is safe to call pipeWriter.CloseWithError(nil)

Reply:
you can simplify the code flow like this:
```
690                        if !mustCache || result.ContentLength < clnt.minSize {                                                                                                                                  
691                                _, err := io.Copy(w, io.LimitReader(result.Body, result.ContentLength))                                                                                                         
692                                if err != nil {                                                                                                                                                                 
693                                        logger.LogIf(context.Background(), err)                                                                                                                                 
694                                        pipeWriter.CloseWithError(err)                                                                                                                                          
695                                }                                                                                                                                                                               
696                                return                                                                                                                                                                          
697                        }                                                                                                                                                                                       
698                        pipeReader, pipeWriter := io.Pipe()                                                                                                                                                     
699                        mw = io.MultiWriter(w, pipeWriter)                                                                                                                                                      
700                                                                                                                                                                                                                
701                        statusCode := result.StatusCode                                                                                                                                                         
702                        w.WriteHeader(statusCode)                                                                                                                                                               
703                                                                                                                                                                                                                
704                        go func() {                                                                                                                                                                             
705                                _, err := clnt.PutObject(clnt.bucket, key,                                                                                                                                      
706                                        io.LimitReader(pipeReader, resultContentLength), resultContentLength,                                                                                                   
707                                        "", "", getPutMetadata(result.Header), nil)                                                                                                                             
708                                if err != nil {                                                                                                                                                                 
709                                        clnt.setOffline()                                                                                                                                                       
710                                        logger.LogIf(context.Background(), err, "Failed to cache object")                                                                                                       
711                                }                                                                                                                                                                               
712                                pipeReader.CloseWithError(err)                                                                                                                                                  
713                        }()                                                                                                                                                                                     
714                                                                                                                                                                                                                
715                        _, err := io.Copy(mw, io.LimitReader(result.Body, result.ContentLength))                                                                                                                
716                        logger.LogIf(context.Background(), err)                                                                                                                                                 
717                        return                                                                                                                                                                                  
```